nomos-sdp/src/ledger.rs

```diff
@@ -0,0 +1,1008 @@
+use std::{
+    collections::{HashMap, hash_map::Entry},
+    error::Error,
+    fmt::Debug,
+    marker::PhantomData,
+};
+
+use async_trait::async_trait;
+
+use crate::{
+    BlockNumber, Declaration, DeclarationId, DeclarationMessage, DeclarationUpdate, EventType,
+    Nonce, ProviderId, ProviderInfo, RewardId, RewardMessage, SdpMessage, ServiceParameters,
+    ServiceType, WithdrawMessage,
+    state::{ProviderState, ProviderStateError},
+};
+
+#[derive(thiserror::Error, Debug)]
+pub enum DeclarationsRepositoryError {
+    #[error("Provider not found: {0:?}")]
+    ProviderNotFound(ProviderId),
+    #[error("Declaration not found: {0:?}")]
+    DeclarationNotFound(DeclarationId),
+    #[error("Duplicate nonce")]
+    DuplicateNonce,
+    #[error(transparent)]
+    Other(Box<dyn Error + Send>),
+}
+
+#[async_trait]
+pub trait DeclarationsRepository {
```

Comment:
Is this intended for the only use of the ledger right? Or should this be a subtrait of the overall defined query API?: https://www.notion.so/Service-Declaration-Protocol-Specification-17b8f96fb65c80c69c2ef55e22e29506?pvs=4#17c8f96fb65c80558624f8892aa55d25

Reply:
You're correct, this is only for sdp ledger use. Ledger only cares about the functions defined in this trait, the query API will probably be implemented by the service or some other entity.